### ISSUE TITLE
Test reproducibility of spaCy training

### ISSUE THREAD

**Issue by FrancescoCasalegno** [BlueBrain/Search#343]
## Context
We have already seen (see [BBS-198](https://bbpteam.epfl.ch/project/issues/browse/BBS-198)) that `torch` training results  (i.e. model weights) are not bitwise reproducible.

On the other hand, `spaCy` training has been reproducible until now. But we have been using `tok2vec` as a backbone, not `transformer`, so can this change the situation?

## Actions
- [x] Test if `spacy` NER training is reproducible when using a `transformer` backbone instead of `tok2vec`.
- [x] If No to the previous question, is the training also not reproducible with a frozen `transformer`.
- [x] If training appears not to be reproducible, ask `spaCy` developers if this is expected—it seems indeed to contrast with the multiple times that "reproducibile" [is mentioned in their docs](https://spacy.io/usage/training).

**Comment by Stannislav** [comment]
Maybe we should also test if reproducibility  holds both on the CPU and the GPU. See also #337

**Comment by jankrepl** [comment]
So far I have looked into training on a CPU. To make the below more readable, let me define 2 concepts of equality

* **md5 equality** - `md5` hashes computed on the contents of 2 files are equal
* **array equality** - (assuming the underlying checkpoint stores some arrays e.g. network weights) when we load the 2 files into  Python the arrays are numerically (!even without tolerance!) equal
 
Note that md5 equality implies array equality but not the other way around! See below a proof by example and what IMO is the source of our problem.

```bash
 python -c "import torch;torch.save(torch.tensor([1., 2.]), 'tensor.pt')" && md5 tensor.pt
```
gives a different `md5` hash each time you call it

```
MD5 (tensor.pt) = c5bc49f7e05cbf019c83c8c30bea364b
MD5 (tensor.pt) = 78cdb7f834c8a1ca79c9611d9f23a7e9
MD5 (tensor.pt) = 7cbbf60f684d542642308202199bf5bf
....
```
* **Non transformer model** - all files within the final trained checkpoint folder are **md5 equal** across different runs
* **Transformer model** -  all files within the final trained checkpoint folder are **md5 equal** across different runs except for the transformers weights `transformer/model/pytorch_model.bin`. However, they are array equal. Which means that the network weights and therefore the predictions are identical.

**Comment by FrancescoCasalegno** [comment]
The following `patch` seems to allow reproducible results.
Note: the only line `patch`ed is 
```python
            obj_key = str(random.randint(0, 10_000_000))
```
and I don't think we can just set `obj_key` to a constant value, since this parameter is used to store each tensor in a separate place...

### Code (`save_tensor.py`)
```python
import io
from unittest.mock import patch
import random

import torch
from torch.serialization import location_tag, normalize_storage_type

random.seed(0)


def patched_save(obj, zip_file, pickle_module, pickle_protocol):
    """See torch.serialization._save"""
    serialized_storages = {}

    def persistent_id(obj):
        # FIXME: the docs say that persistent_id should only return a string
        # but torch store returns tuples. This works only in the binary protocol
        # see
        # https://docs.python.org/2/library/pickle.html#pickling-and-unpickling-external-objects
        # https://github.com/python/cpython/blob/master/Lib/pickle.py#L527-L537
        if torch.is_storage(obj):
            storage_type = normalize_storage_type(type(obj))
            obj_key = str(random.randint(0, 10_000_000))
            location = location_tag(obj)
            serialized_storages[obj_key] = obj

            return ('storage',
                    storage_type,
                    obj_key,
                    location,
                    obj.size())
        return None

    # Write the pickle data for `obj`
    data_buf = io.BytesIO()
    pickler = pickle_module.Pickler(data_buf, protocol=pickle_protocol)
    pickler.persistent_id = persistent_id
    pickler.dump(obj)
    data_value = data_buf.getvalue()
    zip_file.write_record('data.pkl', data_value, len(data_value))

    # Write each tensor to a file named tensor/the_tensor_key in the zip archive
    for key in sorted(serialized_storages.keys()):
        name = f'data/{key}'
        storage = serialized_storages[key]
        # given that we copy things around anyway, we might use storage.cpu()
        # this means to that to get tensors serialized, you need to implement
        # .cpu() on the underlying Storage
        if storage.device.type != 'cpu':
            storage = storage.cpu()
        # Now that it is on the CPU we can directly copy it into the zip file
        num_bytes = storage.size() * storage.element_size()
        zip_file.write_record(name, storage.data_ptr(), num_bytes)


x = torch.tensor([1., 2.])

with patch("torch.serialization._save", patched_save):
    torch.save(x, 'tensor.pt')
```


### Output
```bash
$ python save_tensor.py &&  md5 tensor.pt
MD5 (tensor.pt) = cb85b83c6cd78461f65f237d2e6c947b
```

**Comment by jankrepl** [comment]
@FrancescoCasalegno Very interesting and I will need some time to think about the implementation details. Can you try whether your patch makes the following script reproducible too? IMO it way closer to what is happening in `transformers` (at least I would guess).

```python
from transformers import AutoModel                                                                                                                                              
                                                                                                                                                                                
model = AutoModel.from_pretrained("bert-base-uncased")                                                                                                                          
model.save_pretrained("bert_1")                                                                                                                                   
                                                                                                                                                                                
del model                                                                                                                                                                       
model = AutoModel.from_pretrained("bert-base-uncased")                                                                                                                          
model.save_pretrained("bert_2")       
```

**Comment by FrancescoCasalegno** [comment]
Hi @jankrepl,
The reason why you don't get the same `md5sum` for `bert_1/pytorch_model.bin` and `bert_1/pytorch_model.bin` is that I set
```python
random.seed(0)
```
at the beginning of the script. So to get the same result it would be enough to reset the `seed`.
### Code (`save_model.py`)
```python
from transformers import AutoModel
from save_tensor import patched_save

model = AutoModel.from_pretrained("bert-base-uncased")

random.seed(0)

with patch("torch.serialization._save", patched_save):
    model.save_pretrained("bert_1")

random.seed(0)

with patch("torch.serialization._save", patched_save):
    model.save_pretrained("bert_2")
```
### Output
```bash
$ python save_model.py &&  md5 bert_1/pytorch_model.bin &&  md5 bert_2/pytorch_model.bin
MD5 (bert_1/pytorch_model.bin) = 36aff36ac61f5926f97098ce67d3eeb5
MD5 (bert_2/pytorch_model.bin) = 36aff36ac61f5926f97098ce67d3eeb5
```

**Comment by FrancescoCasalegno** [comment]
As discussed together, in order to patch we will likely have to replace the current call to `spacy train`
https://github.com/BlueBrain/Search/blob/0bb500551b1b7c6f5bb9228335aa4df30a654e9c/data_and_models/pipelines/ner/dvc.yaml#L44-L50
with the execution of a custom script
```yaml
         python spacy_train.py 
         config.cfg 
         --output ../../models/ner/${key}/ 
         --code initialize.py 
         --paths.train ../../annotations/ner/${item.annotation_filename}.train.spacy 
         --paths.dev ../../annotations/ner/${item.annotation_filename}.dev.spacy 
         --paths.base_model ${item.base_model} 
```
where `spacy_train.py` will contain a simple call to [`spacy.cli.train`](https://github.com/explosion/spaCy/blob/49aed683cce4d58baca10e7cb4fe89fbfc209a36/spacy/cli/train.py#L15-L27) wrapped by `random.seed(0)` and `with patch("torch.serialization._save", patched_save):` as shown above.

**Comment by pafonta** [comment]
Hello @FrancescoCasalegno,

> in order to patch we will likely have to replace the current call to `spacy train`

Before going outside the `spacy train` best practices by creating an ad-hoc logic, I would suggest to consider 2 things:

First, the random seed is already fixed in the `config.cfg` (see [here](https://github.com/BlueBrain/Search/blob/0bb500551b1b7c6f5bb9228335aa4df30a654e9c/data_and_models/pipelines/ner/config.cfg#L84) for example). So, it might not be needed to fix it again.

Second, the argument `--code` of `spacy train` lets one import code before launching anything ([docs](https://spacy.io/usage/training#custom-code)). So, the patching of `PyTorch` could happen in the Python file given as value of `--code`.

**Comment by FrancescoCasalegno** [comment]
## Preliminary Summary

These results were obtained using the code  on branch `repro_test`, commit 57f215b771370b09f44f1a3c3ae2d4f016386843.

### Training Reproducibility
<table>
 <tr>
  <td>&nbsp;</td>
  <td><b>CPU</b></td>
  <td><b>GPU</b></td>
 </tr>
 <tr>
  <td><b>Tok2Vec</b></td>
  <td>✅</td>
  <td>❌ <sup>2</sup></td>
 </tr>
 <tr>
  <td><b>Transformer</b></td>
  <td>✅ <sup>1</sup></td>
  <td>❌ <sup>3</sup></td>
 </tr>
</table>

**Notes**
1. Transformers weights match, but the binary files change `md5` sum from run to run.  The reason for this and a fix are discussed [here](https://github.com/BlueBrain/Search/issues/343#issuecomment-829083979).
   - `model-best/transformer/model/pytorch_model.bin`
   - `model-last/transformer/model/pytorch_model.bin`
2. The following files change from run to run. Since metrics also (slightly!) differ, it is likely that the learned weights are really different and the problem is not something like what we discussed at point 1.
   - `model-best/meta.json`
   - `model-last/meta.json`
   - `model-best/ner/model`
   - `model-last/ner/model`
   <details>
          <img width="550" alt="Screenshot 2021-04-30 at 19 54 53" src="https://user-images.githubusercontent.com/17013890/116734937-27cd2b00-a9ee-11eb-9c64-fa19a656351f.png">
          <img width="550" alt="Screenshot 2021-04-30 at 19 55 15" src="https://user-images.githubusercontent.com/17013890/116734943-2ac81b80-a9ee-11eb-86d3-ec1916696fc3.png">
   </details>
3. Both the Transformers weights binaries and the files mentioned at point 2. differ. Unlike point 1., the weights of the Transformers are not equal so it's not just a problem related to how weights are saved to disk.
   - `model-best/transformer/model/pytorch_model.bin`
   - `model-last/transformer/model/pytorch_model.bin` 
   - `model-best/meta.json`
   - `model-last/meta.json`
   - `model-best/ner/model`
   - `model-last/ner/model`
   <details>
          <img width="550" alt="Screenshot 2021-04-30 at 19 58 24" src="https://user-images.githubusercontent.com/17013890/116735187-77abf200-a9ee-11eb-9a8a-a4ff7578e8e1.png">
          <img width="550" alt="Screenshot 2021-04-30 at 19 57 53" src="https://user-images.githubusercontent.com/17013890/116735282-93af9380-a9ee-11eb-8099-353086c12ccc.png">
   </details>
   
### Conclusions
1. On CPU, everything seems to be reproducible. If we use a `Transformer` backbone (which is going to be the case), then we  just need to apply the patch `torch.save` to make sure we have bitwise deterministic binary outputs.
2. On GPU, the training loss seems to be different from run to run even if we don't have a `Transformer` backbone. The metrics `P`, `R`, `F1` seems to match—at least in the tests we run—but the loss term is slightly different and the weights are different (with an average relative difference being about `< 1.e-5`). 
3. As a consequence of the previous point 2., training on CPU vs. GPU give different results for the same settings.

### To Do's
- [ ] Ask `spacy` devs if they are aware of the non-reproducibility of trainings on GPU.
  - [ ] If they provide a way to fix this, ensure that CPU and GPU trainings provide the same results.
- [ ] Ask `dvc` devs if they know about the non-reproducibility of saving a `torch.Tensor`. Are we really the first ones to try to use `dvc` to track a trained `torch` model and realize we need to monkeypatch?
- [ ]  If the monkepatching of `toch.save` proves necessary, check that the current suggested solution allows to load models that function correctly (i.e. they have same predictions on some tests data compared to same models saved w/o the patch).
- [ ] If the monkepatching of `toch.save` proves necessary, try to do it by passing some `--code` during `spacy train` that manipulates the `spacy.cli.train` function itself (see https://github.com/BlueBrain/Search/issues/343#issuecomment-829286577).

**Comment by FrancescoCasalegno** [comment]
## Runtimes

Running on CPU, **training 10 steps takes ~40 s**.

With the default options (`patience = 1600`, `max_epochs = 0`, `max_steps = 20000`), early stopping halts the training after 1,760 steps. So **the total training runtime on CPU is ~2 hours**.

So we wondered whether we really needed to train for this long or if we could set `patience` to a a smaller value. See next section for the results.


## Metrics and Early Stopping
Early stopping is based on the `score`, which defaults to the F1-score. It looks like no significant learning is happening after ~250 epochs (if anything, it even seems to overfit after that). 
❓Should we try to set `patience` to `250` from now on?

See [train_output.txt](https://github.com/BlueBrain/Search/files/6415406/train_output.txt) for the raw data. Loss terms are computed on the `train` set, while scores are on the `dev` set.

<image src="https://user-images.githubusercontent.com/17013890/116891674-4706cb00-ac2f-11eb-9fff-3bc1e4eff38a.png" width="600px">
<image src="https://user-images.githubusercontent.com/17013890/116891122-a3b5b600-ac2e-11eb-824a-14dbab2312bf.png" width="600px">

**Comment by Stannislav** [comment]
Blocked: waiting for progress on external tickets:
- https://github.com/explosion/spaCy/issues/7981

**Comment by FrancescoCasalegno** [comment]
### Update 2021-05-17
- `torch` merged the PR we opened on their code base to fix reproducibility: https://github.com/pytorch/pytorch/pull/57536.
- The latest release as of now was on 2021-03-25 so `pip install --ugprade torch` won't install the reproducibility patch, and we may still have to wait a bit for a new `torch` version with the patch to be released.

**Comment by FrancescoCasalegno** [comment]
## Closing
- https://github.com/BlueBrain/Search/pull/359#issuecomment-848225187 – issues related to `torch` should be now be solved.
- https://github.com/BlueBrain/Search/issues/379 – issues related to GPU are still there, but blocked as we wait for `spacy` maintainers to address them.
